### PR TITLE
feat: 결제 내역 API 및 ErrorCode 수정

### PR DESCRIPTION
**Pull request by hyunaaaj**
### 구현 기능

- 결제 내역 API  잘못된 요청 확인 코드 추가
- ErrorCode 수정

### 구현 코드

- PaymentServiceImpl

`@Override
    @Transactional
    public Long writePayment(PaymentRequestDTO paymentRequestDTO, String token) {
        Long userId = jwtUtil.getUserId(token);
        User user = userRepository.findByUserId(userId)
                .orElseThrow(() -> new CustomException(ErrorCode.USER_NOT_FOUND));

        if (paymentRequestDTO.getPaymentDetail() == null || paymentRequestDTO.getPaymentDetail().isEmpty()) {
            throw new CustomException(ErrorCode.INVALID_PAYMENT_DETAIL1);
        }

        PaymentDetail convertToPaymentDetail;
        try {
            convertToPaymentDetail = PaymentDetail.fromString(paymentRequestDTO.getPaymentDetail());
        } catch (Exception e) {
            throw new CustomException(ErrorCode.INVALID_PAYMENT_DETAIL2);
        }

        if (paymentRequestDTO.getDueDate() == null || paymentRequestDTO.getLastPayment() == null) {
            throw new CustomException(ErrorCode.INVALID_PAYMENT_DETAIL3);
        }

        LocalDate dueDate;
        LocalDate lastPayment;
        try {
            dueDate = LocalDate.parse(paymentRequestDTO.getDueDate());
            lastPayment = LocalDate.parse(paymentRequestDTO.getLastPayment());
        } catch (DateTimeParseException e) {
            throw new CustomException(ErrorCode.INVALID_PAYMENT_DETAIL4);
        }

        Payment payment = Payment.builder()
                .dueDate(dueDate.atStartOfDay())
                .paymentDetail(convertToPaymentDetail)
                .lastPayment(lastPayment.atStartOfDay())
                .cost(paymentRequestDTO.getCost())
                .tradeName(paymentRequestDTO.getTradeName())
                .user(user)
                .build();

        try {
            return paymentRepository.save(payment).getPaymentId();
        } catch (DataIntegrityViolationException e) {
            throw new CustomException(ErrorCode.INVALID_PAYMENT_DETAIL5);
        }
    }
    @Override
    public List<PaymentResponseDTO> listPayments(String token) {
        //유저 정보 가져오기
        Optional<User> user = userRepository.findByUserId(jwtUtil.getUserId(token));
        List<Payment> paymentList = paymentRepository.findAllByUser(user.orElseThrow(
                () -> new CustomException(ErrorCode.USER_NOT_FOUND)
        ));
        if (paymentList.isEmpty()) {
            // 결제 내역이 없을 경우 예외 발생
            throw new CustomException(ErrorCode.PAYMENT_NOT_FOUND);
        }
        return paymentList.stream()
                .map(payment -> PaymentResponseDTO.builder()
                        .paymentId(payment.getPaymentId())
                        .dueDate(payment.getDueDate().toString().substring(0, 10))
                        .paymentDetail(PaymentDetail.toString(payment.getPaymentDetail()))
                        .lastPayment(payment.getLastPayment().toString().substring(0, 10))
                        .cost(payment.getCost())
                        .tradeName(payment.getTradeName())
                        .build())
                .collect(Collectors.toList());
    }

    @Override
    public Long updatePayment(String token, Long paymentId, PaymentRequestDTO paymentRequestDTO) {
        // 유저 정보 가져오기
        User myUser = userRepository.findByUserId(jwtUtil.getUserId(token))
                .orElseThrow(() -> new CustomException(ErrorCode.USER_NOT_FOUND));
        // 결제아이디로 결제 내역 찾기
        Payment findPayment = paymentRepository.findByPaymentId(paymentId)
                .orElse(null);
        if (findPayment == null) {
            throw new CustomException(ErrorCode.PAYMENT_NOT_FOUND);
        }
        // 유저의 지출 내역인지 확인
        if (!Objects.equals(findPayment.getUser().getUserId(), myUser.getUserId())) {
            throw new CustomException(ErrorCode.USER_NOT_CORRECT);
        }
        // 결제 내역 조회
        Payment payment = paymentRepository.findByPaymentId(paymentId)
                .orElseThrow(() -> new CustomException(ErrorCode.PAYMENT_NOT_FOUND));
        // 결제 내역 수정
        Payment updatePayment = findPayment.builder()
                .dueDate(LocalDateTime.parse(paymentRequestDTO.getDueDate() + "T00:00"))
                .paymentDetail(PaymentDetail.fromString(paymentRequestDTO.getPaymentDetail()))
                .lastPayment(LocalDateTime.parse(paymentRequestDTO.getLastPayment() + "T00:00"))
                .cost(paymentRequestDTO.getCost())
                .tradeName(paymentRequestDTO.getTradeName())
                .build();
        // 변경된 결제 내역 저장
        paymentRepository.save(updatePayment);
        // 수정된 결제 내역 ID 반환
        return updatePayment.getPaymentId();
    }

    @Override
    public Long deletePayment(String token, Long paymentId) {
        Optional<User> user = userRepository.findByUserId(jwtUtil.getUserId(token));
        User currentUser = user.orElseThrow(() ->  new CustomException(ErrorCode.USER_NOT_FOUND));

        Payment payment = paymentRepository.findById(paymentId)
                .orElseThrow(() -> new CustomException(ErrorCode.PAYMENT_NOT_FOUND));

        // 결제 내역 삭제
        paymentRepository.delete(payment);
        return paymentId;
    }`

- ErrorCode

`INVALID_PAYMENT_DETAIL(400, "유효하지 않은 결제 상세 정보입니다."),
    INVALID_PAYMENT_DETAIL1(400, "결제 상세 정보가 비어 있습니다."),
    INVALID_PAYMENT_DETAIL2(400, "결제 상세 정보를 변환할 수 없습니다."),
    INVALID_PAYMENT_DETAIL3(400, "결제 예정일 또는 마지막 결제일이 비어 있습니다."),
    INVALID_PAYMENT_DETAIL4(400, "결제 날짜 형식이 유효하지 않습니다."),
    INVALID_PAYMENT_DETAIL5(400, "결제 데이터를 저장할 수 없습니다."),
    // 400 BAD_REQUEST 잘못된 요청
    INVALID_PAYMENT_PARAMETER(400, "잘못된 결제 정보입니다."),
    PAYMENT_NOT_FOUND(404, "존재하지 않는 결제 내역입니다."),
    USER_NOT_FOUND_FOR_PAYMENT(404, "결제 정보를 찾을 수 없는 사용자입니다."),
    // 409 CONFLICT 결제 내역 충돌
    PAYMENT_CONFLICT(409, "결제 내역이 이미 존재합니다."),
`

### CODE REVIEW

File: src/main/java/com/WealthTracker/demo/DTO/PaymentRequestDTO.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+@AllArgsConstructor
+@NoArgsConstructor
+
+public class PaymentRequestDTO {
+    //  결제 ID
+    private Long paymentId;
+    // 결제 상세 내용
+    private String paymentDetail;
+    // 결제 예정일
+    private String dueDate;
```
Comment (pjs1710) [inline src/main/java/com/WealthTracker/demo/DTO/PaymentRequestDTO.java:19]: 여기서 `dueDate`와 `lastPayment`를 `String Type`으로 지정하신 이유가 있을까요?

File: src/main/java/com/WealthTracker/demo/DTO/PaymentResponseDTO.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+@AllArgsConstructor
+@NoArgsConstructor
+
+public class PaymentResponseDTO {
+    //  결제 ID
+    private Long paymentId;
+    // 결제 상세 내용
+    private String paymentDetail;
+    // 결제 예정일
+    private String dueDate;
```
Comment (pjs1710) [inline src/main/java/com/WealthTracker/demo/DTO/PaymentResponseDTO.java:19]: 이 부분도 위와 마찬가지의 질문입니다

File: src/main/java/com/WealthTracker/demo/constants/ErrorCode.java [hunk 3]
```diff
@@ -6,6 +6,19 @@
 @Getter
 @AllArgsConstructor
 public enum ErrorCode {
+    INVALID_PAYMENT_DETAIL(400, "유효하지 않은 결제 상세 정보입니다."),
```
Comment (pjs1710) [inline src/main/java/com/WealthTracker/demo/constants/ErrorCode.java:9]: 이 부분은 제가 현아님께 오류에 대한 로그를 찾아주고자 `ENUM`의 이름을 1,2,3,4,5 이렇게 설정했던 겁니다!
각 오류에 해당하는 이름을 알아보기 쉽도록 다시 설정해주는 것이 좋아보여요.
ex) `PAYMENT_DETAIL_EMPTY(400, "결제 상세 정보가 비어 있습니다.")`

File: src/main/java/com/WealthTracker/demo/controller/PaymentController.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+import org.springframework.http.HttpStatus;
+import org.springframework.http.HttpStatusCode;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/api")
```
Comment (pjs1710) [inline src/main/java/com/WealthTracker/demo/controller/PaymentController.java:28]: 각 결제 내역에 API를 확인해보시면 `/payment`가 공통으로 들어가있어서 이 부분도 `@RequestMapping`으로 빼주시면 가독성이 더 좋을거에요 🙂

File: src/main/java/com/WealthTracker/demo/controller/PaymentController.java [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        try {
+            // 지출 내역 수정 메서드 호출
+            paymentService.updatePayment(token,paymentId,paymentRequestDTO);
+            // 성공 메시지 반환
+            ReturnCodeDTO returnCodeDTO = ReturnCodeDTO.builder()
+                    .status(200)
+                    .message("수정 성공")
+                    .build();
+            return new ResponseEntity<>(returnCodeDTO, HttpStatus.OK);
+        }catch (CustomException ex){
```
Comment (pjs1710) [inline src/main/java/com/WealthTracker/demo/controller/PaymentController.java:118]: 결제 내역에 대한 CRUD 구현하느라 고생하셨습니다! 예외 처리도 잘 해주셨네요

File: src/main/java/com/WealthTracker/demo/domain/Payment.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+@AllArgsConstructor(access=AccessLevel.PROTECTED)
+@NoArgsConstructor
+
+public class Payment {
+    @Id
+    @GeneratedValue(strategy=GenerationType.IDENTITY)
+    @Column(unique = true)
+    private Long paymentId;
+
+    private LocalDateTime dueDate;
```
Comment (pjs1710) [inline src/main/java/com/WealthTracker/demo/domain/Payment.java:23]: 첫 질문과 동일하게 `dueDate` 필드는 `DTO`에선 `String Type`인데 `Entity`에서는 `LocalDateTime`으로 parsing을 한번 더 해야하는 번거로움이 있을 거에요.

**Review by pjs1710 — COMMENTED** [review]
결제 내역에 대한 CRUD는 잘 작성하셨는데 결제 예정 일자가 되면 사용자의 지출이 반영되도록 구현하셨을까요?

File: src/main/resources/application.properties
Comment (pjs1710) [inline src/main/resources/application.properties:1]: 해당 파일은 외부에 노출되어선 안되는 서버에 중요한 파일을 기록하는 곳이라고 생각하면 됩니다. 민감한 정보인 `Key` 값들은 삭제해주신 뒤 올려주세요 😎

**Review by tkv00 — COMMENTED** [review]
수고하셨습니다! 우선 추가 구현해야할 API 구현하시고 제목을 hotfix가 아닌 feat으로 수정 부탁드립니다.
또한 저희가 프로젝트 시작할 때 정한 `branch commit`규칙을 반드시 지켜주세요!

**Comment by tkv00** [comment]
추가로 PR 내용이 보면 코드가 깨집니다..수정부탁드릴게요.

**Pull request merged**